Mathlib/NumberTheory/NumberField/ClassNumber.lean

```diff
@@ -40,51 +41,103 @@ variable {K}
+  refine ⟨1, fun C ↦ ?_⟩
+  obtain ⟨I, rfl, hI⟩ := exists_ideal_in_class_of_norm_le C
+  simpa [← ClassGroup.mk0_eq_one_iff] using h hI
+
+theorem _root_.RingOfIntegers.isPrincipalIdealRing_of_isPrincipal_of_norm_le_of_isPrime
+    (h : ∀ ⦃I : (Ideal (𝓞 K))⁰⦄, (I : Ideal (𝓞 K)).IsPrime →
+      absNorm (I : Ideal (𝓞 K)) ≤ (4 / π) ^ nrComplexPlaces K *
+      ((finrank ℚ K)! / (finrank ℚ K) ^ (finrank ℚ K) * √|discr K|) →
+      Submodule.IsPrincipal (I : Ideal (𝓞 K))) :
+    IsPrincipalIdealRing (𝓞 K) := by
+  refine RingOfIntegers.isPrincipalIdealRing_of_isPrincipal_of_norm_le (fun I hI ↦ ?_)
+  rw [← mem_isPrincipalSubmonoid_iff,
+    ← prod_normalizedFactors_eq_self (nonZeroDivisors.coe_ne_zero I)]
+  refine Submonoid.multiset_prod_mem _ _ (fun J hJ ↦ mem_isPrincipalSubmonoid_iff.mp ?_)
+  by_cases hJ0 : J = 0
+  · simpa [hJ0] using bot_isPrincipal
+  rw [← Subtype.coe_mk J (mem_nonZeroDivisors_of_ne_zero hJ0)]
+  refine h (((mem_normalizedFactors_iff (nonZeroDivisors.coe_ne_zero I)).mp hJ).1) ?_
+  exact (cast_le.mpr <| le_of_dvd (absNorm_pos_of_nonZeroDivisors I) <|
+    absNorm_dvd_absNorm_of_le (le_of_dvd (UniqueFactorizationMonoid.dvd_of_mem_normalizedFactors
+    hJ))).trans hI
+
+theorem _root_.RingOfIntegers.isPrincipalIdealRing_of_isPrincipal_of_mem_primesOver
+    (h : ∀ ⦃p : ℕ⦄, p.Prime → p ≤ (4 / π) ^ nrComplexPlaces K *
+      ((finrank ℚ K)! / (finrank ℚ K) ^ (finrank ℚ K) * √|discr K|) →
+      ∀ (I : Ideal (𝓞 K)), I ∈ Ideal.primesOver (span {(p : ℤ)}) (𝓞 K) →
+      Submodule.IsPrincipal I) :
+    IsPrincipalIdealRing (𝓞 K) := by
+  refine RingOfIntegers.isPrincipalIdealRing_of_isPrincipal_of_norm_le_of_isPrime <|
+    fun P hP hPNorm ↦ ?_
+  obtain ⟨p, hp⟩ := IsPrincipalIdealRing.principal <| under ℤ P.1
+  have hp0 : p ≠ 0 := fun h ↦ nonZeroDivisors.coe_ne_zero P <|
+    eq_bot_of_comap_eq_bot (R := ℤ) <| by simpa only [hp, submodule_span_eq, span_singleton_eq_bot]
+  have hpprime := (span_singleton_prime hp0).mp
+  rw [← submodule_span_eq, ← hp] at hpprime
+  have hle : algebraMap ℤ (𝓞 K) p ∈ P.1 := (mem_of_liesOver P.1 (under ℤ P.1) p).mp <|
+    hp ▸ Submodule.mem_span_singleton_self p
+  refine h (Int.prime_iff_natAbs_prime.mp (hpprime (hP.under _))) ?_ _ ⟨hP, ?_⟩
+  · refine le_trans (cast_le.mpr <| Nat.le_of_dvd ?_ (Int.ofNat_dvd_right.mp ?_)) hPNorm
+    · exact Nat.pos_of_ne_zero <| fun h ↦ nonZeroDivisors.coe_ne_zero P <| absNorm_eq_zero_iff.mp h
+    suffices (Algebra.norm ℤ (algebraMap ℤ (𝓞 K) p)) = p ^ (Module.finrank ℚ K) by
+      obtain ⟨i, -, hi⟩ := (dvd_prime_pow (hpprime (IsPrime.comap _)) _).mp
+        (this ▸ absNorm_dvd_norm_of_mem hle)
+      refine dvd_trans (dvd_pow_self p (fun h ↦ hP.ne_top (absNorm_eq_one_iff.mp ?_))) hi.dvd'
+      simp only [h, pow_zero, associated_one_iff_isUnit] at hi
+      exact ZMod.eq_one_of_isUnit_natCast hi
+    exact Rat.intCast_injective (by simp [Algebra.coe_norm_int, ← Algebra.norm_algebraMap])
+  · convert over_under P.1
```

Comment:
`convert` really needed? Again, not at my VSCode, but seems a pity in an otherwise neat proof.

Reply:
Fixed.